lucene/misc/src/test/org/apache/lucene/misc/store/TestDirectIODirectory.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.misc.store;
+
+import com.carrotsearch.randomizedtesting.LifecycleScope;
+import com.carrotsearch.randomizedtesting.RandomizedTest;
+import org.apache.lucene.store.*;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.apache.lucene.misc.store.DirectIODirectory.DEFAULT_MIN_BYTES_DIRECT;
+
+public class TestDirectIODirectory extends BaseDirectoryTestCase {
+  public void testWriteReadWithDirectIO() throws IOException {
+    try(Directory dir = getDirectory(RandomizedTest.newTempDir(LifecycleScope.TEST))) {
+      final long blockSize = Files.getFileStore(createTempFile()).getBlockSize();
+      final long minBytesDirect = Double.valueOf(Math.ceil(DEFAULT_MIN_BYTES_DIRECT / blockSize)).longValue() *
+                                    blockSize;
+      // Need to worry about overflows here?
+      final int writtenByteLength = Math.toIntExact(minBytesDirect);
+
+      MergeInfo mergeInfo = new MergeInfo(1000, Integer.MAX_VALUE, true, 1);
+      final IOContext context = new IOContext(mergeInfo);
+
+      IndexOutput indexOutput = dir.createOutput("test", context);
+      indexOutput.writeBytes(new byte[writtenByteLength], 0, writtenByteLength);
+      IndexInput indexInput = dir.openInput("test", context);
+
+      assertEquals("The length of bytes read should equal to written", writtenByteLength, indexInput.length());
+
+      indexOutput.close();
+      indexInput.close();
+    }
+  }
+
+  @Override
+  protected Directory getDirectory(Path path) throws IOException {
+    Directory delegate = FSDirectory.open(path);
```

Comment:
We also have `LuceneTestCase.newFSDirectory(path)`, which adds some randomization. I'd suggest to use this method instead.

Reply:
I verified: `BaseDirectoryTestCase` uses [newIOContext(Random)](https://lucene.apache.org/core/8_7_0/test-framework/org/apache/lucene/util/LuceneTestCase.html#newIOContext-java.util.Random-). So we have the correct randomization and our directory is used from time to time.

Reply:
I just gave this a try, but looks like it would consistently fail some test cases in `BaseDirectoryTestCase`:

1. With plain replacement of `FSDirectory.open(path)` with `LuceneTestCase.newFSDirectory(path)`, this would fail  `BaseDirectoryTestCase.testCreateOutputForExistingFile`. The issue is `DirectIODirectory` delegates to `MockDirectoryWrapper.createOutput` for file creation, but `FSDirectory.deleteFile` for file deletion. This cause `MockDirectoryWrapper.createdFiles` variable to not be updated properly for deletion, and thus test failed with exception `FileAlreadyExistsException`
2. When I also tried to delegate `DirectIODirectory.deleteFile` to `MockDirectoryWrapper.deleteFile` with 
  ```
  @Override
  public void deleteFile(String name) throws IOException {
    delegate.deleteFile(name);
  }
  ```

   `BaseDirectoryTestCase.testCreateOutputForExistingFile` will now pass, but `BaseDirectoryTestCase.testPendingDeletions`  will start to fail with `FileAlreadyExistsException`, with tempDir folder containing a bunch of random files not deleted due to `VirusCheckingFS` rejecting deletion. I'm still debugging this but it may take some time.

Reply:
I've figured it out. Looks like more methods in `DirectIODirectory` need to be delegated. Could you please take a look at the latest commit, and let me know if if looks good?